Assets/Tests/PlayMode/EOSSessionsManagerTests.cs

```diff
@@ -0,0 +1,654 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+namespace PlayEveryWare.EpicOnlineServices.Tests.IntegrationTests
+{
+    using Epic.OnlineServices;
+    using Epic.OnlineServices.Sessions;
+    using NUnit.Framework;
+    using EpicOnlineServices;
+    using System.Collections;
+    using UnityEngine;
+    using UnityEngine.TestTools;
+    using PlayEveryWare.EpicOnlineServices.Samples;
+    using System;
+    using UnityEngine.Events;
+    using System.Text;
+    using System.Collections.Generic;
+    using System.Text.RegularExpressions;
+
+    public class EOSSessionsManagerTests : EOSTestBase
+    {
+        protected EOSSessionsManager ManagerInstance { get; set; } = null;
+
+        /// <summary>
+        /// The Epic Online Services Game Services back end takes some fuzzy amount of time to process changes.
+        /// For example, a created game won't appear immediately in search results, even if the creation operation succeeded.
+        /// Tests should use this constant amount of wait time before running online queries.
+        /// </summary>
+        const float SecondsWaitAfterSessionModification = 6f;
+
+        static bool[] TestCasesForCreateSessionAndTogglePublicVisibility = { true, false };
+
+        [OneTimeSetUp]
+        public void OneTimeSetUpEOSSessionsManager()
+        {
+            ManagerInstance = new EOSSessionsManager();
+            Debug.Log($"{nameof(OneTimeSetUpEOSSessionsManager)}");
+        }
+
+        [SetUp]
+        public void SetUpLogIn()
+        {
+            ManagerInstance.OnLoggedIn();
+            Debug.Log($"{nameof(SetUpLogIn)}");
+        }
+
+        [UnityTearDown]
+        public IEnumerator TearDownLogOut()
+        {
+            // If there are any Sessions that are remaining locally, they all need to be left
+            // This may not successfully destroy all Sessions, but should at least attempt
+            if (ManagerInstance.GetCurrentSessions().Count > 0)
+            {
+                ManagerInstance.LeaveAllSessions();
+
+                yield return new WaitUntilDone(60f, () =>
+                {
+                    return ManagerInstance.GetCurrentSessions().Count == 0;
+                });
+
+                yield return new WaitForSeconds(SecondsWaitAfterSessionModification);
+
+                ManagerInstance.OnLoggedOut();
+            }
+        }
+
+        /// <summary>
+        /// Essential test of the most basic functions of the Manager.
+        /// Creates a Session, observes that it was created successfully, leaves the Session.
+        /// </summary>
+        [UnityTest]
+        public IEnumerator CreateSessionLeaveSession()
```

Comment:
Looking at several of the tests defined in this class, I am seeing that there is a not-insignificant overlap between the functionality of each test; particularly the "CreateSession" portion of the test. Could we perhaps extract the portion of each test that is responsible for session creation so that we can consolidate that into a central location for better maintainability?

Reply:
How do you think we should go about doing this? The `CreateSession` utility function feels like it's already doing this as best as can be done.

Reply:
I agree - I failed to notice that. I think there is still room for greater reduction of duplicated code - but not so significant that it ought to warrant a delay in approval. Put another way: the value of the tests as written outweigh the value of over-engineering a solution to reduce the code footprint to its smallest possible form.